compiler/rustc_middle/src/middle/region.rs

```diff
@@ -234,6 +234,15 @@ pub struct ScopeTree {
     /// escape into 'static and should have no local cleanup scope.
     rvalue_scopes: FxHashMap<hir::ItemLocalId, Option<Scope>>,
 
+    /// Sometimes destruction scopes are too overarching in the sense that
+    /// all temporaries are dropped while it may be the case that
+    /// only a subset of temporaries is requested to be dropped.
+    /// For instance, `x[y.z()]` indexing operation can discard the temporaries
+    /// generated in evaluating `y.z()`.
```

Comment:
> For instance, `x[y.z()]` indexing operation can discard the temporaries generated in evaluating `y.z()`.

I'll admit, I still need to reload my mental cache here, but: Is this actually true? Aren't there cases where the value produced by `y.z()` may hold references to those temporaries, so you'd need to keep them alive across the evaluation of the indexing operation?

I'll do some investigation to double-check my understanding here, and see if I can convince myself that you are correct.

Reply:
(While talking to niko about this, he notes the example of `mutex.lock().something()` as a potentially problematic example...)

Reply:
I think this was initially bending my mind as well. Let me try to elaborate. So the indexing operation, either through [`Index`](https://doc.rust-lang.org/stable/core/ops/trait.Index.html#tymethod.index) or [`IndexMut`](https://doc.rust-lang.org/stable/core/ops/trait.IndexMut.html#tymethod.index_mut), produces borrows with lifetimes that must outlives `self`.
Let us take `Index` as an example. In this example, we attempt to allow indexing into `A` with a temporary reference `&()`, producing a value of type `&B<'_>`. The result of this indexing will need the temporary index `&()` to be kept alive
```rust
use std::ops::Index;

struct A;

struct B<'a>(&'a ());

impl<'a> B<'a> {
    fn foo(&self) {}
}

impl<'a> Index<&'a ()> for A {
    type Output = B<'a>;
    fn index<'s>(&'s self, idx: &'a ()) -> &'s Self::Output {
        unimplemented!() // (1)
    }
}

fn main() {
    let a = A;
    a[&()].foo();
}
```
At `(1)`, we are expected to construct a borrow with a lifetime `'s`. However, the `Index` trait definition does allow us to impose `Self::Output: 's` or any outliving relation between `'a` and `'s` in this scope. Overall, this means that if the index, which is `'_ ()` in this example, has a lifetime `'a`, `'s` must outlives `'a`.
<details>
<summary><i>A side track</i></summary>
I think there should be an implicit bound of `Self::Output: 's` to be satisfied, but the current compiler does not complain as long as the return value is `!`.
</details>
So the remaining question is in which way we shall honor this contract. Indeed, this may have been upheld so far through enlarging the terminating scope.

Regarding the example from @nikomatsakis, let us example the following elaborated code.
```rust
use std::ops::{Index, IndexMut, DerefMut};
use std::cell::Cell;
use std::sync::Mutex;

struct A<'a>(Cell<&'a mut ()>);

impl<'a> Index<&'a mut ()> for A<'a> {
    type Output = ();
    fn index(&self, idx: &'a mut ()) -> &Self::Output {
        &*self.0.replace(idx)
    }
}

impl<'a> IndexMut<&'a mut ()> for A<'a> {
    fn index_mut(&mut self, idx: &'a mut ()) -> &mut Self::Output {
        self.0.replace(idx)
    }
}

struct B(());

impl B {
    fn new() -> B {
        B(())
    }
    fn foo(&mut self) -> &mut () {
        &mut self.0
    }
}

fn main() {
    let t = &mut ();
    let mut a = A(Cell::new(t));
    let i = Mutex::new(());
    a[i.lock().unwrap().deref_mut()] = (); // <- temporary value is freed at the end of this statement
    a[B::new().foo()] = (); // <- same reason
    a[&mut ()] = ();
}
```
The expression `a[i.lock().unwrap().deref_mut()] = ()` is currently rejected because the temporary mutex guard `mutex.lock()` is preserved only until the end of this assignment.

In any case, I am also experimenting ways to produce examples that requiring living temporaries after index to be extra sure.

Reply:
Okay.

It looks like you are making an argument based on the lifetimes embedded in the `Idx` type of `trait Index<Idx>`.

I think I can see the reasoning there; its something like "the definition of the `Index` trait: `pub trait Index<Idx: ?Sized> { type Output: ?Sized; fn index(&self, index: Idx) -> &Self::Output; }`, will force any free lifetimes associated with the `Idx` to be bound at the scope of the `impl<'a, 'b, ...> Index<...> for Concrete<...>` for any concrete type `Concrete<...>`."

My follow-up question then is: Is there a case where the `Idx` type doesn't have any lifetimes in it, but the scope of temporaries of the expression computing `Idx` are nonetheless significant (and where the change suggested here could break code)?

My current *suspicion* is that the most natural cases where that could arise is that it could only happen in unsafe code that is *incorrect*. (I.e., unsafe code that *should* be using lifetimes to express the constraints between the temporaries and the dynamic extent of the produced value, but is sidestepping doing so, and in the process opening itself up to bugs).
 * I think we have precedent for making changes like that, but I also think we need to be careful whenever we do so.

Reply:
(In case its not clear, I'm going to poke around a bit more and see if I can make my concerns here concrete. My current thinking is that you may indeed be safe in the assertions you are making here, but I want to make sure I understand the edge cases really well.)